### ISSUE TITLE
Create a proxy app

### ISSUE THREAD

**Issue by Tratcher** [microsoft/reverse-proxy#247]
We've proposed shipping YARP in three formats:
- A library - https://www.nuget.org/packages/Microsoft.ReverseProxy/
- A template - https://github.com/microsoft/reverse-proxy/issues/77
- A pre-built application exe (this issue)

The pre-built app should:
- Read from a config file
- Cover basic scenarios
- Support a variety of platforms

# Scenarios
- Config based routes, clusters, & backends
- Remote route config? E.g. load from a database, service fabric, k8?
- Dashboards
- Health checks for the proxy itself.
- Static files?
- Auth?
- Multi-instance?
  - Data protection (Session Affinity)
- Single-file exe with linking: https://github.com/microsoft/reverse-proxy/issues/7

# Config
The library already supports loading routes from a config file. However, an app will need to expose additional config for the scenarios described above.

# Platforms
- Windows, Mac, Linux
- Builds for both runtime dependent and standalone (servicing?)

# Acquisition
- Microsoft download center?

**Comment by b0** [comment]
Is there use for a docker image with this pre-built application exe as another shipping format?

**Comment by Tratcher** [comment]
@b0 as an ingress controller (https://github.com/microsoft/reverse-proxy/issues/200) or as a simplified deployment model?

Packaging this exe up into a pre-built container would be one step to consider beyond the development of the exe itself (e.g. additional work item). I'd primarily wonder about the configuration story in that case.

**Comment by b0** [comment]
@Tratcher I find both usable. It's possible to host containers in Azure App Service too(OpenShift, etc...) so a plain docker image could have value. As of configuration story for that kind of container I find docker volumes and a plain json file YARP already support more then sufficient...  Hope this have sense for you?

**Comment by Tratcher** [comment]
@b0 Good to know. We may split containerizing out into a separate issue when we get around to working on the exe scenario.

**Comment by samsp-msft** [comment]
Is this about having a pre-built container image for YARP, or having a docker file so that customers can build their own images?
For configuration does the config file have to live next to the binary, or is there a way that config can be directed to be loaded from another well-known location that could then be mapped to being stored elsewhere by docker.

**Comment by b0** [comment]
It's about pre-built image with everything ready for running. Configuration file(routes, destinations, etc...) doesn't need to be inside image nor next to binary but can be provided via docker volumes - pretty standard way of doing things. Volumes are the way to load config from other location and not keeping it next to binary.

**Comment by samsp-msft** [comment]
From the comments it seems like creating a docker image is also popular. As this issue had the comments, I have changed this to be creating a docker image, and forked a new issue #261 to track the stand alone exe.

**Comment by lukemcdo** [comment]
This would just be a small layer on top of the existing ASP.NET images, right?

**Comment by Tratcher** [comment]
@lukemcdo correct, just adding the YARP library and a standard Startup + config.

**Comment by damianog** [comment]
@Tratcher is it possible to configure the endopoint to map the reverse proxy on the standard ports 80,443 and route request from another port like 81 to a classic MVC controller route?

`.MapWhen(p => p.Request.Host.Port == 81, etc.. `

What I want to do is a yarp service with a simple mvc backend with authentication from which to configure routes and clusters and save them on a static file or a sqlite db.

Thanks in advance.

**Comment by PureKrome** [comment]
Hi Team - with YARP 1.0 now RTM'd, is there plans to review this issue (an official docker image) for vNext ?

Sure, the docker image can only service a number of scenario's, but it would be a pretty large suite, IMO. Could really help with competing with other RP's.

**Comment by samsp-msft** [comment]
We can, but I don't think that really shows off the benefits of YARP - being able to customize the way it handles the non-simple cases - such as authentication, dynamic config etc. If you don't need that, you won't get a great deal of benefit out of YARP compared to other RPs.

**Comment by PureKrome** [comment]
@samsp-msft : ok, maybe I could ask the question in another way. What's the most benefit one person could get from a docker'd instance of YARP via providing a config file? Surely this list of Out-of-the-box with a config file can handle a large _common_ number of scenario's?

Like for example, I want to use YARP instead of Traefik:
- Given these routes, direct them to these backend destination/server(s)/"clusters". 
- Log metrics to this-metrics-server (so I can see stuff)
- expose the dashboard so I can see stuff

and this is (i would have thought) a pretty common _entry level_ scenario. 

Am I misunderstanding what YARP is, then?

**Comment by dpbevin** [comment]
Does this feature cover vanilla YARP *and* the Kubernetes YARP?

I'd like to see an image for the Kubernetes flavoured YARP.

**Comment by mguinness** [comment]
It looks like docker images [YARP](https://hub.docker.com/r/microsoft/dotnet-nightly-yarp) & [.NET Reverse Proxy (Preview)](https://hub.docker.com/r/microsoft/dotnet-nightly-reverse-proxy) are already available on Docker Hub.

For anyone looking for a secure web proxy (https and webauthn), take a look at [YarpWebAuthn](https://hub.docker.com/r/mguinness/yarpwebauthn) image.
